### PR TITLE
Add sigusr2 pprof file dump support

### PR DESCRIPTION
**Pull request by jasonodonnell**
This adds support to dump `pprof` outputs to the filesystem when `VAULT_PPROF_SIGUSR2_DUMP` set on the Vault process. This is very useful in situations where HTTP handlers are unresponsive and we cannot use the `sys/pprof` endpoints. I put this behind an environment variable because files are generated each time there's a SIGUSR2 signal to avoid unintentionally generating a lot of files.

### CODE REVIEW

**Review by hghaf099 — APPROVED** [review]
Looks good to me! Great improvement!

File: command/server.go [hunk 1]
```diff
@@ -1641,39 +1641,86 @@ func (c *ServerCommand) Run(args []string) int {
...
-						continue
-					}
+			if os.Getenv("VAULT_PPROF_SIGUSR2_DUMP") != "" {
+				dir, err := os.MkdirTemp("", "pprof")
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				fGoRoutine, err := os.Create(fmt.Sprintf("%s/goroutines", dir))
```
Comment (ncabatoff) [inline command/server.go:1651]: It looks like you're dropping the existing "stacktrace" output, which was the comprehensive text version of the stacks, and replacing it with this pprof version of the stacks, which may not include everything.  You're also removing the ability to specify a path for where the files should go. Could you retain the existing functionality please?

File: command/server.go [hunk 2]
```diff
@@ -1641,39 +1641,86 @@ func (c *ServerCommand) Run(args []string) int {
...
+					continue
+				}
+
+				err = pprof.Lookup("goroutine").WriteTo(fGoRoutine, 0)
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				fHeap, err := os.Create(fmt.Sprintf("%s/heap", dir))
```
Comment (ncabatoff) [inline command/server.go:1663]: Use filepath.Join over fmt.Sprintf for this kind of thing.

File: command/server.go [hunk 3]
```diff
@@ -1641,39 +1641,86 @@ func (c *ServerCommand) Run(args []string) int {
...
+					continue
+				}
+
+				err = pprof.Lookup("allocs").WriteTo(fAllocs, 0)
+				if err != nil {
+					c.UI.Error(err.Error())
+					continue
+				}
+
+				fMutex, err := os.Create(fmt.Sprintf("%s/mutex", dir))
```
Comment (ncabatoff) [inline command/server.go:1687]: IIRC the mutex and block pprofs won't do anything unless runtime tunables are set first, maybe we shouldn't bother unless that's done?

File: command/server.go [hunk 4]
```diff
@@ -1641,39 +1641,86 @@ func (c *ServerCommand) Run(args []string) int {
...
-						continue
-					}
-					dir = path
-				} else {
-					dir, err = os.MkdirTemp("", "vault-stacktrace")
-					if err != nil {
-						c.logger.Error("Could not create temporary directory for stacktrace", "error", err)
-						continue
-					}
+			if os.Getenv("VAULT_PPROF_SIGUSR2_DUMP") != "" {
```
Comment (ncabatoff) [inline command/server.go:1644]: Ideally, I'd like to see the output format be a debug bundle that can be consumed by debug-helper, and for it to include metrics.  Needn't be part of this PR, just thought I'd mention it so you could consider it.

Whether or not you take that suggestion, another idea to consider: allow an env var (possibly  VAULT_PPROF_SIGUSR2_DUMP itself) to specify an optional duration, and when that's provided, collect cpu profile info for that amount of time, including it as another output file.

File: command/server.go
Comment (peteski22) [inline command/server.go:1]: Just a thought, could this be put into a library package to make it more easily testable?

**Review by elliesterner — APPROVED** [review]

**Pull request closed without merging**